### PR TITLE
Improve perf of relationship snapshot

### PR DESCRIPTION
**Pull request by ajcvickers**
This is done by calculating indexes for the properties that need to be snapshotted and using an array sidecar to store the snapshot.

Perf improvement is about 15%, with also about 10% improvement in memory usage.

### CODE REVIEW

File: src/EntityFramework.Core/ChangeTracking/Internal/RelationshipsSnapshot.cs [hunk 1]
```diff
@@ -1,33 +1,28 @@
...
-        {
-            var entityType = entry.EntityType;
+        protected override int Index(IPropertyBase property) => property.GetRelationshipIndex();
 
-            return entityType.GetKeys().SelectMany(k => k.Properties)
-                .Concat(entityType.GetForeignKeys().SelectMany(fk => fk.Properties))
-                .Distinct()
-                .Concat<IPropertyBase>(entityType.GetNavigations());
+        protected override void ThrowInvalidIndexException(IPropertyBase property)
+        {
```
Comment (anpete) [inline src/EntityFramework.Core/ChangeTracking/Internal/RelationshipsSnapshot.cs]: Should property be used here?

File: src/EntityFramework.Core/Metadata/Internal/PropertyIndexes.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.Data.Entity.Metadata.Internal
+{
+    public class PropertyIndexes
```
Comment (anpete) [inline src/EntityFramework.Core/Metadata/Internal/PropertyIndexes.cs]: Consider a struct

**Comment by anpete** [comment]
:shipit:

File: src/EntityFramework.Core/ChangeTracking/Internal/RelationshipsSnapshot.cs (hunk 1 above)
Comment (ajcvickers) [inline src/EntityFramework.Core/ChangeTracking/Internal/RelationshipsSnapshot.cs]: This should never get called now, outside of an application error. I'm planning to remove in another checkin.

File: src/EntityFramework.Core/Metadata/Internal/PropertyIndexes.cs (hunk 2 above)
Comment (ajcvickers) [inline src/EntityFramework.Core/Metadata/Internal/PropertyIndexes.cs]: I don't think it's going to help making it a struct, but I'll look into it.

**Pull request merged**